test/docker-e2e/e2e_test.go

```diff
@@ -171,3 +187,256 @@ func (s *CelestiaTestSuite) WaitForSync(ctx context.Context, statusClient rpccli
 		}
 	}
 }
+
+// CheckLiveness validates that all validators proposed blocks and no nodes halted.
+// Automatically waits for sufficient blocks (3 per validator minimum) if needed.
+//
+// Upgrade-agnostic: can be called before/after upgrades or spanning the entire period.
+// Call at the end of E2E tests to validate network health.
+func (s *CelestiaTestSuite) CheckLiveness(
+	ctx context.Context,
+	chain tastoratypes.Chain,
+	startHeight int64,
+) error {
+	rpcClient, err := chain.GetNodes()[0].GetRPCClient()
+	if err != nil {
+		return fmt.Errorf("failed to get RPC client: %w", err)
+	}
+
+	endHeight, err := s.ensureMinimumBlocks(ctx, chain, rpcClient, startHeight)
+	if err != nil {
+		return fmt.Errorf("failed to ensure minimum blocks: %w", err)
+	}
+
+	startValidators, endValidators, err := s.fetchValidatorSets(ctx, rpcClient, startHeight, endHeight)
+	if err != nil {
+		return fmt.Errorf("failed to fetch validator sets: %w", err)
+	}
+
+	proposers, err := s.fetchProposerAddresses(ctx, rpcClient, startHeight, endHeight)
+	if err != nil {
+		return fmt.Errorf("failed to fetch proposer addresses: %w", err)
+	}
+
+	if err := s.validateAllValidatorsProposed(startValidators, endValidators, proposers, startHeight, endHeight); err != nil {
+		return err
+	}
+
+	if err := s.validateNodesNotHalted(ctx, chain, endHeight); err != nil {
+		return err
+	}
+
+	s.T().Logf("✅ Liveness check passed: all validators proposed blocks and all nodes reached height %d", endHeight)
+	return nil
+}
+
+// ensureMinimumBlocks waits for additional blocks if necessary to meet the minimum
+// requirement of defaultBlocksPerValidator (3) blocks per validator since startHeight.
+func (s *CelestiaTestSuite) ensureMinimumBlocks(
+	ctx context.Context,
+	chain tastoratypes.Chain,
+	rpcClient rpcclient.Client,
+	startHeight int64,
+) (int64, error) {
+	currentStatus, err := rpcClient.Status(ctx)
+	if err != nil {
+		return 0, fmt.Errorf("failed to get current status: %w", err)
+	}
+	currentHeight := currentStatus.SyncInfo.LatestBlockHeight
+	blocksProduced := currentHeight - startHeight
+
+	// Get validator count to calculate minimum required blocks
+	validators, err := rpcClient.Validators(ctx, &currentHeight, nil, nil)
+	if err != nil {
+		return 0, fmt.Errorf("failed to get validators: %w", err)
+	}
+	numValidators := len(validators.Validators)
+	minBlocksRequired := int64(numValidators * defaultBlocksPerValidator)
+
+	if blocksProduced >= minBlocksRequired {
+		s.T().Logf(blockRequirementMetLogFormat,
+			blocksProduced, minBlocksRequired, numValidators, defaultBlocksPerValidator)
```

Comment:
[uber nit] I find it more difficult to read code when there are random newlines introduced. This applies to many other instances in this PR

```suggestion
		s.T().Logf(blockRequirementMetLogFormat, blocksProduced, minBlocksRequired, numValidators, defaultBlocksPerValidator)
```